### DISCUSSION TITLE
"out var" scope

### DISCUSSION THREAD

**Discussion by yaakov-h** [dotnet/csharplang#3893] [General] [answered]
Is there an _intentional_ difference in the scope of an `out var`-declared variable depending on the structure it is defined in?

I'm seeing a strange difference where if I declare a variable in `if (Foo(out var thing))` then the scope leaks, but if it is declared in as `while (Foo(out var thing))` then the scope does not leak.

i.e.:

```csharp
async Task DoWork()
{
    while (!queue.TryTake(out var nextFile))
    {
        await Task.Delay(TimeSpan.FromMilliseconds(100));
    }

    MyBigBallOf(nextFile); // error CS0103: The name 'nextFile' does not exist in the current context
}
```

Sharplab playground [here](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQxAN0QSwGYE8A+AAgEwCMAsAFCEAMABIaQHQDCA9gDYdwDGMmbAHZRWQngFcECOIJgBuKrQbMAKgAtpEACaZBAcwXV6jAKyHFAZgbE6LKgG86TulWe2IgnnC4R+QlWwA1jIAymySXnR8UHQAvHRacNgQ4hzyrs7snpLSsgBCEHoAPCEwCLp6AHx0AI7icPVxCUkpaYZuGU6EABwMAGx0ACJsAOpsCIEAFACUnXT2c24A7mqY3HSTAIR1DXBMKgi4KhDBk+EwdGgQCHSCcAAeMABia3DTs5RubgufX3+EAE5+kxBt4ILhJipMGA4CEAA4eJhPBBsMAAWTWHEwUF4Qi0UEmpBoNHe7T+dAAvlRFs40bg8pg9AUuAB5bCTO6PF7caZkpxU35OOY9fpDUbjQLEGZzH7kug4Dbber1faHY6nc6Xa63B7PV7vGlOWVy5yA4GgjjgyHQ2EIwRIlHozHY3GCfGE4mkw2U6mCr50hlMiCs9mcvU8vk+v3C0gDQgAFjoAcZzI4bMmjHosAQH2+zjmAopQA===).

I would very much like for this to work without an error.

**Comment by HaloFour** [comment, marked as answer]
Yes, pattern variables and `out` declarations within a `while` loop condition do not leak to the outer scope.  This was an explicit decision based on the propensity that a capturing lambda within the loop body would enclose over the single leaked variable scope and not the value of the variable within the loop.  That was a common source of confusion and bugs with capturing in `foreach` loops to the extent that the team was willing to accept a breaking change to the language to resolve, so they didn't want to repeat that mistake.

https://github.com/dotnet/csharplang/blob/master/meetings/2016/LDM-2016-11-30.md#scope-of-while-condition-expression-variables
https://github.com/dotnet/roslyn/issues/15529
